pallets/farming/src/lib.rs

```diff
@@ -0,0 +1,930 @@
+            pid: PoolId,
+        },
+        FarmingPoolEdited {
+            pid: PoolId,
+        },
+        Charged {
+            who: AccountIdOf<T>,
+            pid: PoolId,
+            rewards: Vec<(CurrencyIdOf<T>, BalanceOf<T>)>,
+        },
+        Deposited {
+            who: AccountIdOf<T>,
+            pid: PoolId,
+            add_value: BalanceOf<T>,
+            gauge_info: Option<(BalanceOf<T>, BlockNumberFor<T>)>,
+        },
+        Withdrawn {
+            who: AccountIdOf<T>,
+            pid: PoolId,
+            remove_value: Option<BalanceOf<T>>,
+        },
+        Claimed {
+            who: AccountIdOf<T>,
+            pid: PoolId,
+        },
+        WithdrawClaimed {
+            who: AccountIdOf<T>,
+            pid: PoolId,
+        },
+        GaugeWithdrawn {
+            who: AccountIdOf<T>,
+            gid: PoolId,
+        },
+        AllForceGaugeClaimed {
+            gid: PoolId,
+        },
+        PartiallyForceGaugeClaimed {
+            gid: PoolId,
+        },
+        AllRetired {
+            pid: PoolId,
+        },
+        PartiallyRetired {
+            pid: PoolId,
+        },
+        RetireLimitSet {
+            limit: u32,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Pool not exist
+        PoolDoesNotExist,
+        /// Gauge pool not exist
+        GaugePoolNotExist,
+        /// Gauge info not exist
+        GaugeInfoNotExist,
+        /// Share info is not exist
+        ShareInfoNotExists,
+        /// Pool state is invalid
+        InvalidPoolState,
+        /// Can not claim gauge
+        LastGaugeNotClaim,
+        /// claim_limit_time exceeded
+        CanNotClaim,
+        /// gauge pool max_block exceeded
+        GaugeMaxBlockOverflow,
+        /// withdraw_limit_time exceeded
+        WithdrawLimitCountExceeded,
+        /// Can not deposit to pool yet
+        CanNotDeposit,
+    }
+
+    /// The next farming pool id.
+    #[pallet::storage]
+    #[pallet::getter(fn pool_next_id)]
+    pub type PoolNextId<T: Config> = StorageValue<_, PoolId, ValueQuery>;
+
+    /// The next gauge farming pool id.
+    #[pallet::storage]
+    #[pallet::getter(fn gauge_pool_next_id)]
+    pub type GaugePoolNextId<T: Config> = StorageValue<_, PoolId, ValueQuery>;
+
+    /// The retire limit of one operation when retire farming pool.
+    #[pallet::storage]
+    #[pallet::getter(fn retire_limit)]
+    pub type RetireLimit<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+    /// Record reward pool info.
+    ///
+    /// map PoolId => PoolInfo
+    #[pallet::storage]
+    #[pallet::getter(fn pool_infos)]
+    pub type PoolInfos<T: Config> = StorageMap<_, Twox64Concat, PoolId, PoolInfoOf<T>>;
+
+    /// Record gauge farming pool info.
+    ///
+    /// map PoolId => GaugePoolInfo
+    #[pallet::storage]
+    #[pallet::getter(fn gauge_pool_infos)]
+    pub type GaugePoolInfos<T: Config> = StorageMap<_, Twox64Concat, PoolId, GaugePoolInfoOf<T>>;
```

Comment:
Feel like this should be a bounded structure as there are for loops in the code that can explode
@Apokalip https://github.com/Manta-Network/Manta/issues/624

Reply:
@ghzlatarev Not sure how you would bound this without loosing the key feature of the StorageMap, do you have example code?

Reply:
actually nvm, i forget this is a map.